notebooks-auth/resources/user_test_cowbird_jupyter.ipynb

```diff
@@ -0,0 +1,172 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #3.    public_wps_outputs_filepath = os.path.join(notebook_dir, "public/wps_outputs/weaver/public/test_public_file.txt")
</pre><p> </p><p>I'm surprised that this path has the double <code>public</code> nesting, but the following user path output does not.</p><p> </p><p>On server host, we should have something like:</p><ul><li>/data/wps_outputs/weaver/public/{jobs_uuid}/{output_id}/test_public_file.txt</li><li>/data/wps_outputs/weaver/users/{test_user_id}/{jobs_uuid}/{output_id}/test_user_file.txt</li></ul><p> </p><p>If I remember correctly, those should be mapped to something like:</p><ul><li>{notebook_dir}/public/wps_outputs/weaver/{jobs_uuid}/{output_id}/test_public_file.txt</li><li>{notebook_dir}/wps_outputs/{jobs_uuid}/{output_id}/test_user_file.txt</li></ul><p> </p><p>Are the different paths in this notebook caused by an incorrect parsing from Cowbird file-system mount, or is it simply because the sample paths specified here are not based on real examples? If the latter, I would suggest adjusting the path to represent "expected" locations to avoid causing confusion, even if outcomes from the test should remain the same.</p><p> </p><p> </p><p> </p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/131/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Reply:
<p>On the host, the source file is created at <code>/data/wps_outputs/weaver/public/test_public_file.txt</code>.</p><p> </p><p>If the source file does not follow the pattern <code>&lt;wps_outputs_dir&gt;/&lt;bird_name&gt;/users/&lt;user_id&gt;/...</code> then it is considered a public file. </p><p> </p><p>Cowbird then finds the source file path relative to the WPS outputs dir (example here <code>weaver/public/test_public_file.txt</code> and creates a hardlink at the path <code>/data/user_workspaces/public/wps_outputs/&lt;relative_path&gt;</code> Replace the relative path here with the value subpath found just above, for the result <code>&lt;workdir&gt;/public/wps_outputs/weaver/public/test_public_file.txt</code>  </p><p> </p><p>The path <code>/data/user_workspaces/public</code> is mounted on <code>/notebook_dir/public</code> on JupyterLab.</p><p> </p><p>Basically all the public WPS outputs are placed in a <code>public</code> directory, and Cowbird is not made to customize the relative path that is used. I think this approach was used because each bird might have different structures, or there could be other files or folders in the bird directory that could be considered public.</p><p> </p><p>So I think the test content here is right considering the design choice made in Cowbird. If we want to change this, we would have to update Cowbird.</p><p> </p>

Reply:
Sorry if I wasn't clear. 
The host path and the hardlink patterns (the prefixes) are correct. 
It is the remaining `test_public_file.txt` that looks odd. 
After `/weaver/public`, it should always be `{job_uuid}/{output_id}/test_public_file.txt` or `{job_uuid}.xml` / `{job_uuid}.log`.
There is no condition where the WPS output file would be directly under `/weaver/public`, so that might create a false expectation from the test if it was used as an example. 
The security aspect of those files, however how nested they are, should remain the same as how you already test them.

Reply:
Oh, I thought the comment was about the fact that `public` was found 2 times in the url.

If I understand, I would just have to add the `{job_uuid}/{output_id}` subdirs to better represent an actual real use case.

Reply:
Exact 👍